Global limit for MSQ controller tasks implemented

nozjkoitop opened the pull request:
# Improvements

Implemented a way to limit on the number of query controller tasks running at the same time. This limit specifies what percentage or amount of task slots can be allocated to query controllers. If the limit is reached, the tasks would wait for resources instead of potentially blocking the execution of other tasks (and failing after a timeout).

### Rationale

 There is no mechanism in Druid to prevent the cluster from being overloaded with controller tasks. Currently, it could cause a significant slowdown in processing and may lead to temporary deadlock situations.

### Introduced new configuration options

* `druid.indexer.queue.controllerTaskSlotRatio` - optional value which defines the proportion of available task slots that can be allocated to msq controller tasks. This is a floating-point value between 0 and 1. Defaults to null.
* `druid.indexer.queue.maxControllerTaskSlots` - optional value which specifies the maximum number of task slots that can be allocated to controller tasks. This is an integer value that provides a hard limit on the number of task slots available for msq controller tasks. Defaults to null.

This PR has:

- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

asdf2014 commented on docs/configuration/index.md:
```diff
@@ -1124,6 +1124,8 @@ These Overlord static configurations can be defined in the `overlord/runtime.pro
 |`druid.indexer.queue.restartDelay`|Sleep this long when Overlord queue management throws an exception before trying again.|`PT30S`|
 |`druid.indexer.queue.storageSyncRate`|Sync Overlord state this often with an underlying task persistence mechanism.|`PT1M`|
 |`druid.indexer.queue.maxTaskPayloadSize`|Maximum allowed size in bytes of a single task payload accepted by the Overlord.|none (allow all task payload sizes)|
+|`druid.indexer.queue.controllerTaskSlotRatio`|Optional value which defines the proportion of available task slots that can be allocated to `query_controller` tasks. This is a floating-point value between 0 and 1. |`null`|
```
```suggestion
|`druid.indexer.queue.controllerTaskSlotRatio`|Optional value which defines the proportion of available task slots that can be allocated to `query_controller` tasks. This is a floating-point value between 0 and 1.|`null`|
```

asdf2014 reviewed (commented):
Remove useless tail

nozjkoitop commented:
> Remove useless tail

Done, thanks

kfaraz reviewed (commented):
> Currently, it could cause a significant slowdown in processing and may lead to temporary deadlock situations.

@nozjkoitop , rather than proceeding with a limit, I think we should try to figure out what is causing the slowdown in processing and/or the deadlock. Can you elaborate on this?

Once we have done an analysis of exactly what goes wrong when we have too many controller tasks and we have decided to impose a limit, it should not be done through the `TaskQueue` as done in this PR. Instead, it should be similar to the implementation of `parallelIndexTaskSlotRatio` and the config should most likely live in `WorkerTaskRunnerConfig`.

https://github.com/apache/druid/blob/73ff9f9047d69d55bf880e629efeec07ce92e528/indexing-service/src/main/java/org/apache/druid/indexing/overlord/config/WorkerTaskRunnerConfig.java#L37-L50

cryptoe commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java:
```diff
@@ -440,7 +442,7 @@ private void manageInternalCritical(
             continue;
           }
-          if (taskIsReady) {
+          if (taskIsReady && !isControllerTaskLimitReached(task.getType(), true)) {
```
Should It just be limited to MSQ controllers or other job types as well. Maybe Take a json as a input ? where key is the taskType and the value is the limit/float. 

Also what is the user behavior  if the task is pending for launch due to limit. 
* Can we communicate why the task is not launching to the user.
* If the cluster is totally starved, does the controller get timed out eventually and removed from queue?

cryptoe reviewed (commented):
I think this is a nice feature to have. Left some comments. 
Thanks @nozjkoitop for taking this up.

nozjkoitop commented:
Thanks @kfaraz, @cryptoe  for your comments 
The most trivial deadlock scenario occurs when we queue a group of controller tasks but don't have available task slots for actual workers. This results in tasks hanging and eventually timing out. Thanks for highlighting the WorkerTaskRunnerConfig, it seems like the great place for this configuration. I've updated the behavior and merged it with `parallelIndexTaskSlotRatio`, now it's more flexible, also I've added the logging to inform the user why the task is Pending

kfaraz commented on docs/configuration/index.md:
```diff
@@ -1134,7 +1134,7 @@ The following configs only apply if the Overlord is running in remote mode. For
 |--------|-----------|-------|
 |`druid.indexer.runner.taskAssignmentTimeout`|How long to wait after a task has been assigned to a Middle Manager before throwing an error.|`PT5M`|
 |`druid.indexer.runner.minWorkerVersion`|The minimum Middle Manager version to send tasks to. The version number is a string. This affects the expected behavior during certain operations like comparison against `druid.worker.version`. Specifically, the version comparison follows dictionary order. Use ISO8601 date format for the version to accommodate date comparisons. |"0"|
-| `druid.indexer.runner.parallelIndexTaskSlotRatio`| The ratio of task slots available for parallel indexing supervisor tasks per worker. The specified value must be in the range `[0, 1]`. |1|
```
We cannot change the existing property as there are already people using it.
Please add a new property that can take precedence over the existing property.
You can call it `taskTypeToSlotRatio` or simply `taskSlotRatios`.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -56,7 +60,7 @@ public class ImmutableWorkerInfo
   public ImmutableWorkerInfo(
       @JsonProperty("worker") Worker worker,
       @JsonProperty("currCapacityUsed") int currCapacityUsed,
-      @JsonProperty("currParallelIndexCapacityUsed") int currParallelIndexCapacityUsed,
```
Please do not remove existing fields.

nozjkoitop replied:
Removal rollbacked

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/config/WorkerTaskRunnerConfig.java:
```diff
@@ -20,32 +20,66 @@
   @JsonProperty
-  private double parallelIndexTaskSlotRatio = 1;
+  @JsonDeserialize(using = CustomJobTypeLimitsDeserializer.class)
+  private Map<String, Number> customJobTypeLimits = new HashMap<>();
```
Why do we need a custom deserializer?
Can't we just have a map from String to Double?

nozjkoitop replied:
It was added mostly for validation as the idea is to have double / integer values to have an option to specify not only the ratio but also a limit

kfaraz reviewed (changes requested):
We need to ensure that we do not remove any of the existing properties or config fields.

cryptoe commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -57,6 +63,7 @@ public ImmutableWorkerInfo(
       @JsonProperty("worker") Worker worker,
       @JsonProperty("currCapacityUsed") int currCapacityUsed,
       @JsonProperty("currParallelIndexCapacityUsed") int currParallelIndexCapacityUsed,
+      @JsonProperty("currTypeSpecificCapacityUsed") Map<String, Integer> typeSpecificCapacityMap,
```
This should be nullable no ?

nozjkoitop replied:
You're right, thanks

cryptoe commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -225,6 +253,89 @@ private int getWorkerParallelIndexCapacity(double parallelIndexTaskSlotRatio)
     return workerParallelIndexCapacity;
   }
 
+  public boolean canRunTask(Task task, Map<String, Number> taskLimits)
```
Can you please add java docs for this method.

nozjkoitop replied:
Done

cryptoe commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ZkWorker.java:
```diff
@@ -167,6 +167,29 @@ private int getCurrParallelIndexCapacityUsed(Map<String, TaskAnnouncement> tasks
   }
 
+  @JsonProperty("currTypeSpecificCapacityUsed")
+  public Map<String, Integer> getCurrTypeSpecificCapacityUsed()
```
I thought we had deprecated Zk based runner in favour of http. 
@kfaraz Does this change still make sense ?

kfaraz replied:
Yes, ZK-based task runner is deprecated and we should not support the new feature with ZK.

cryptoe commented on docs/configuration/index.md:
```diff
@@ -1135,6 +1135,7 @@ The following configs only apply if the Overlord is running in remote mode. For
 |`druid.indexer.runner.taskAssignmentTimeout`|How long to wait after a task has been assigned to a Middle Manager before throwing an error.|`PT5M`|
 |`druid.indexer.runner.minWorkerVersion`|The minimum Middle Manager version to send tasks to. The version number is a string. This affects the expected behavior during certain operations like comparison against `druid.worker.version`. Specifically, the version comparison follows dictionary order. Use ISO8601 date format for the version to accommodate date comparisons. |"0"|
 | `druid.indexer.runner.parallelIndexTaskSlotRatio`| The ratio of task slots available for parallel indexing supervisor tasks per worker. The specified value must be in the range `[0, 1]`. |1|
+|`druid.indexer.runner.taskSlotLimits`| A map where each key is a task type, and the corresponding value represents the limit on the number of task slots that a task of that type can occupy on a worker. The key is a `String` that specifies the task type. The value can either be a Double or Integer. A `Double` in the range [0, 1], representing a ratio of the available task slots that tasks of this ...
```
Could you please provide an example as well ?

cryptoe replied:
How does this interact with compaction slots ?

nozjkoitop replied:
Example added! Good catch on the compaction slots. I'll need to test that, but based on the code, it looks like compaction slots availability will be checked twice (in the duty, and during the worker selection) if they're included in the `taskSlotLimits` map. If there's a conflict, some tasks might end up in a pending state for a while.

nozjkoitop replied:
Confirmed that the number of submitted tasks is linked to the compaction task slot limits. However, execution might be delayed if the custom limit is smaller than the one set for compaction.
![image](https://github.com/user-attachments/assets/71a7f10e-b23f-4282-80bd-f07a794a13ab)
For example here,  `taskSlotsMax = 3`, but in the overlord configuration, I have `druid.indexer.runner.taskSlotLimits={"compact": 2}`.

kfaraz replied:
I think the limit on compaction tasks (or kill tasks for that matter) should not be a concern.
This is a runtime property, typically controlled by an admin.
So, if an admin wants to restrict the number of concurrent compaction tasks, it is fair to honor that irrespective of the value of `compactionTaskSlotRatio` or `maxCompactionTaskSlots` set in the coordinator dynamic configs.

We just need to call it out clearly in the release notes and the docs of the new property.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -225,6 +253,114 @@ private int getWorkerParallelIndexCapacity(double parallelIndexTaskSlotRatio)
+   * Determines if a specific task can be executed on the worker based on
+   * various capacity, custom limits, and availability conditions.
+   * <p>
+   * This method checks:
```
```suggestion
   * This method returns true only if:
```

nozjkoitop replied:
Updated

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -225,6 +253,114 @@ private int getWorkerParallelIndexCapacity(double parallelIndexTaskSlotRatio)
+   * <p>
+   * This method checks:
+   * <ul>
+   *   <li>Whether the worker has sufficient capacity to handle the task.</li>
```
```suggestion
   *   <li>The worker has sufficient capacity to handle the task.</li>
```

nozjkoitop replied:
Done

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -225,6 +253,114 @@ private int getWorkerParallelIndexCapacity(double parallelIndexTaskSlotRatio)
+   *         the task type does not exceed custom limits, and the task's availability group is available.
+   *         Returns {@code false} otherwise.
+   */
+  public boolean canRunTask(Task task, Map<String, Number> taskLimits)
```
We don't need to have two separate `canRunTask` methods. You can just pass `WorkerTaskRunnerConfig` instead of `parallelIndexTaskSlotRatio` into the existing `canRunTask` method.

Once we do that, some of the new private methods will go away, and simplify the change set.

nozjkoitop replied:
Merged canRunTask methods

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -260,12 +399,24 @@ public boolean equals(Object o)
+  public Map<String, Integer> incrementTypeSpecificCapacity(String type, int capacityToAdd)
+  {
+    Map<String, Integer> result = new HashMap<>(typeSpecificCapacityMap);
+    if (result.containsKey(type)) {
+      result.put(type, result.get(type) + capacityToAdd);
+    } else {
+      result.put(type, capacityToAdd);
+    }
```
maybe simplified with `result.merge()`.

nozjkoitop replied:
Nice catch, thanks!

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -57,6 +63,7 @@ public ImmutableWorkerInfo(
       @JsonProperty("worker") Worker worker,
       @JsonProperty("currCapacityUsed") int currCapacityUsed,
       @JsonProperty("currParallelIndexCapacityUsed") int currParallelIndexCapacityUsed,
+      @JsonProperty("currTypeSpecificCapacityUsed") @Nullable Map<String, Integer> typeSpecificCapacityMap,
```
Please use the same name for the argument, the JSON serialized field and the class member for readability.

Also, this name is a little ambiguous, use something else like `currCapacityUsedByTaskType`.

nozjkoitop replied:
Updated

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/config/WorkerTaskRunnerConfig.java:
```diff
@@ -48,4 +57,46 @@ public double getParallelIndexTaskSlotRatio()
   }
+
+  /**
+   * The `taskSlotLimits` configuration is a map where each key is a task type,
+   * and the corresponding value represents the limit on the number of task slots
+   * that a task of that type can occupy on a worker.
+   * <p>
+   * The key is a `String` that specifies the task type.
+   * The value can either be a Double or Integer:
+   * <p>
+   * 1. A `Double` in the range [0, 1], representing a ratio of the available task slots
+   * that tasks of this type can occupy. For example, a value of 0.5 means that tasks
+   * of this type can occupy up to 50% of the task slots on a worker.
+   * A value of 0 means that tasks of this type can occupy no slots (i.e., they are effectively disabled).
+   * A value of 1.0 means no restriction, allowing tasks of this type to occupy all available slots.
+   * <p>
+   * 2. An `Integer` that is greater than or equal to 0, representing an absolute limit
+   * on the number of task slots that tasks of this type can occupy. For example, a value of 5
+   * means that tasks of this type can occupy up to 5 task slots on a worker.
+   * <p>
+   * If a task type is not present in the `taskSlotLimits` map, there is no restriction
+   * on the number of task slots it can occupy, meaning it can use all available slots.
+   * <p>
+   * Example:
+   * <p>
+   * taskSlotLimits = {
+   * "index_parallel": 0.5,  // 'index_parallel' can occupy up to 50% of task slots
+   * "query_controller": 3     // 'query_controller' can occupy up to 3 task slots
+   * }
+   * <p>
+   * This configuration allows for granular control over the allocation of task slots
+   * based on the specific needs of different task types, helping to prevent any one type
+   * of task from monopolizing worker resources and reducing the risk of deadlocks.
+   *
+   * @return A map where the key is the task type (`String`), and the value is either a `Double` (0 to 1)
+   * representing the ratio of task slots available for that type, or an `Integer` (>= 0)
+   * representing the absolute limit of task slots for that type. If a task type is absent,
+   * it is not limited in terms of the number of task slots it can occupy.
+   */
```
Please try to simplify and shorten this javadoc retaining only the necessary information in a concise manner.

nozjkoitop replied:
Split that into two blocks, one for each getter and simplified

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/util/TaskSlotLimitsDeserializer.java:
```diff
@@ -0,0 +1,70 @@
+import java.util.Iterator;
+import java.util.Map;
+
+public class TaskSlotLimitsDeserializer extends JsonDeserializer<Map<String, Number>>
```
This should not be needed once task slot ratio and task slot limits are separated.

nozjkoitop replied:
Removed

kfaraz commented on indexing-service/src/test/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfoTest.java:
```diff
@@ -201,6 +211,7 @@ public void testEqualsAndSerde()
         ),
         3,
         0,
+        new HashMap<>(),
```
Use `Collections.emptyMap()` instead.

nozjkoitop replied:
Updated

kfaraz reviewed (changes requested):
Thanks for the quick response, @nozjkoitop !

I haven't reviewed the actual logic in `canRunTaskBasedOnCustomLimit` yet as I feel it will get simplified once the other feedback is addressed.

nozjkoitop commented on docs/configuration/index.md:
```diff
@@ -1135,6 +1135,7 @@ The following configs only apply if the Overlord is running in remote mode. For
 |`druid.indexer.runner.taskAssignmentTimeout`|How long to wait after a task has been assigned to a Middle Manager before throwing an error.|`PT5M`|
 |`druid.indexer.runner.minWorkerVersion`|The minimum Middle Manager version to send tasks to. The version number is a string. This affects the expected behavior during certain operations like comparison against `druid.worker.version`. Specifically, the version comparison follows dictionary order. Use ISO8601 date format for the version to accommodate date comparisons. |"0"|
 | `druid.indexer.runner.parallelIndexTaskSlotRatio`| The ratio of task slots available for parallel indexing supervisor tasks per worker. The specified value must be in the range `[0, 1]`. |1|
+|`druid.indexer.runner.taskSlotLimits`| A map where each key is a task type, and the corresponding value represents the limit on the number of task slots that a task of that type can occupy on a worker. The key is a `String` that specifies the task type. The value can either be a Double or Integer. A `Double` in the range [0, 1], representing a ratio of the available task slots that tasks of this ...
```
I've slit the properties into separate ones. It should be per worker limit I'd say

nozjkoitop commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -225,6 +253,114 @@ private int getWorkerParallelIndexCapacity(double parallelIndexTaskSlotRatio)
+   *                   If the task type is not present in this map, the task can use all available slots.
+   * @return {@code true} if the task can run, meaning the worker has sufficient capacity,
+   *         the task type does not exceed custom limits, and the task's availability group is available.
+   *         Returns {@code false} otherwise.
```
Removed

nozjkoitop commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/config/WorkerTaskRunnerConfig.java:
```diff
@@ -29,6 +34,10 @@ public class WorkerTaskRunnerConfig
 
+  @JsonProperty
+  @JsonDeserialize(using = TaskSlotLimitsDeserializer.class)
+  private Map<String, Number> taskSlotLimits = new HashMap<>();
```
Separated

nozjkoitop commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -42,9 +46,11 @@
 @PublicApi
 public class ImmutableWorkerInfo
 {
+  private static final Logger logger = new Logger(ImmutableWorkerInfo.class);
```
Logger removed, but tbh i don't see the reason why it shouldn't be in this class as we're making the decisions here, should the canRunTask logic be moved into some utility class where we can log what's going on?

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -225,6 +270,79 @@ private int getWorkerParallelIndexCapacity(double parallelIndexTaskSlotRatio)
     return workerParallelIndexCapacity;
   }
 
+  private boolean hasSufficientWorkerCapacity(Task task)
+  {
+    int capacityRemaining = worker.getCapacity() - getCurrCapacityUsed();
+    int requiredCapacity = task.getTaskResource().getRequiredCapacity();
+    return capacityRemaining >= requiredCapacity;
+  }
+
+  private boolean isAvailabilityGroupAvailable(Task task)
+  {
+    return !getAvailabilityGroups().contains(task.getTaskResource().getAvailabilityGroup());
+  }
```
These two methods are not really needed. We can just continue with the original code where these checks were performed inline in the method `canRunTask`.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/config/WorkerTaskRunnerConfig.java:
```diff
@@ -48,4 +56,36 @@ public double getParallelIndexTaskSlotRatio()
+  }
+
+  /**
+   * Returns a map where each key is a task type (`String`), and the value is a `Double` which should be in the
+   * range [0, 1], representing the ratio of available task slots that tasks of this type can occupy.
+   * <p>
+   * This ratio defines the proportion of total task slots a task type can use, calculated as `ratio * totalSlots`.
+   * <p>
+   * If both a ratio and an absolute limit (from {@link #getTaskSlotLimits()}) are specified for the same task type,
+   * the effective limit will be the smaller of the two.
+   *
+   * @return A map of task types with their corresponding slot ratios.
```
```suggestion
   * Map from task type to task slot ratio for that type per worker.
   * A value of 0 for a task type indicates that the task type cannot occupy any slots.
   * A value of 1 indicates that the task type may take up all available slots of a worker if available. 
```

nozjkoitop replied:
Added, thanks

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -260,12 +381,20 @@ public boolean equals(Object o)
              : that.blacklistedUntil != null);
   }
 
+  public Map<String, Integer> incrementTypeSpecificCapacity(String type, int capacityToAdd)
```
```suggestion
  public Map<String, Integer> incrementCurrCapacityUsedByTaskType(String taskType, int capacityToAdd)
```

nozjkoitop replied:
Added

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -195,13 +222,31 @@ public boolean isValidVersion(String minVersion)
   }
 
+
+  /**
+   * To be removed in future as it could be fully covered with the limits map.
+   */
```
I don't think this is needed, especially since we have not deprecated the old property yet.

nozjkoitop replied:
Removed

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -225,6 +270,79 @@ private int getWorkerParallelIndexCapacity(double parallelIndexTaskSlotRatio)
+    return absoluteLimit != null ? absoluteLimit : ratioBasedLimit;
+  }
+
+  private boolean hasCapacityBasedOnLimit(int limit, int currentCapacityUsed, int requiredCapacity)
```
This method is not needed, it can be inlined.

nozjkoitop replied:
Done, thanks for notice

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -225,6 +270,79 @@ private int getWorkerParallelIndexCapacity(double parallelIndexTaskSlotRatio)
+    return !getAvailabilityGroups().contains(task.getTaskResource().getAvailabilityGroup());
+  }
+
+  private boolean canRunTaskBasedOnCustomLimit(Task task, Map<String, Integer> limitsMap, Map<String, Double> ratiosMap)
```
This method will become simpler once we remove the absolute limits.

nozjkoitop replied:
Simplified

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -225,6 +270,79 @@ private int getWorkerParallelIndexCapacity(double parallelIndexTaskSlotRatio)
+
+    // Validate the ratio limit if present
+    if (ratioLimit != null) {
+      Preconditions.checkArgument(ratioLimit >= 0.0 && ratioLimit <= 1.0,
+                                  "Ratio for task %s must be between 0.0 and 1.0 inclusive.", taskType
```
This validation should happen in `WorkerTaskRunnerConfig` itself.

nozjkoitop replied:
Moved to the config using javax.validation

kfaraz reviewed (commented):
Left some more comments.

nozjkoitop commented on docs/configuration/index.md:
```diff
@@ -1135,6 +1135,8 @@ The following configs only apply if the Overlord is running in remote mode. For
 |`druid.indexer.runner.taskAssignmentTimeout`|How long to wait after a task has been assigned to a Middle Manager before throwing an error.|`PT5M`|
 |`druid.indexer.runner.minWorkerVersion`|The minimum Middle Manager version to send tasks to. The version number is a string. This affects the expected behavior during certain operations like comparison against `druid.worker.version`. Specifically, the version comparison follows dictionary order. Use ISO8601 date format for the version to accommodate date comparisons. |"0"|
 | `druid.indexer.runner.parallelIndexTaskSlotRatio`| The ratio of task slots available for parallel indexing supervisor tasks per worker. The specified value must be in the range `[0, 1]`. |1|
+|`druid.indexer.runner.taskSlotLimits`|A map where each key is a task type (`String`), and the corresponding value represents the absolute limit on the number of task slots that tasks of this type can occupy on a worker. The value is an `Integer` that is greater than or equal to 0. For example, a value of 5 means that tasks of this type can occupy up to 5 task slots on a worker. If both absolute  ...
```
Absolute limits removed

nozjkoitop commented on docs/configuration/index.md:
```diff
@@ -1135,6 +1135,8 @@ The following configs only apply if the Overlord is running in remote mode. For
 |`druid.indexer.runner.minWorkerVersion`|The minimum Middle Manager version to send tasks to. The version number is a string. This affects the expected behavior during certain operations like comparison against `druid.worker.version`. Specifically, the version comparison follows dictionary order. Use ISO8601 date format for the version to accommodate date comparisons. |"0"|
 | `druid.indexer.runner.parallelIndexTaskSlotRatio`| The ratio of task slots available for parallel indexing supervisor tasks per worker. The specified value must be in the range `[0, 1]`. |1|
+|`druid.indexer.runner.taskSlotLimits`|A map where each key is a task type (`String`), and the corresponding value represents the absolute limit on the number of task slots that tasks of this type can occupy on a worker. The value is an `Integer` that is greater than or equal to 0. For example, a value of 5 means that tasks of this type can occupy up to 5 task slots on a worker. If both absolute  ...
+|`druid.indexer.runner.taskSlotRatios`|A map where each key is a task type (`String`), and the corresponding value is a `Double` which should be in the range [0, 1], representing the ratio of available task slots that tasks of this type can occupy. This ratio defines the proportion of total task slots a task type can use, calculated as `ratio * totalSlots`. If both absolute and ratio limits are s ...
```
Added, thanks

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -195,21 +221,24 @@ public boolean isValidVersion(String minVersion)
+   * Determines if a specific task can be executed on the worker based on
+   * various capacity, custom limits, and availability conditions.
+   * <p>
+   * returns true only if:
```
```suggestion
   * Returns true only if:
```

nozjkoitop replied:
Done

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -195,21 +221,24 @@ public boolean isValidVersion(String minVersion)
+   * returns true only if:
+   * <ul>
+   *   <li>The worker has sufficient capacity to handle the task.</li>
+   *   <li>The task is of the parallel index type and can operate within the permitted ratio of slots designated for that task type.</li>
+   *   <li>The the task can run under custom-defined limits for its type,
+   *   such as a maximum number of tasks allowed or a ratio of slots the task type can occupy.</li>
```
```suggestion
   *   <li>Task slot ratio for this task type has not been exhausted on this worker.</li>
```

nozjkoitop replied:
Updated

kfaraz reviewed (commented):
Final comments, rest looks good.

nozjkoitop commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/config/WorkerTaskRunnerConfig.java:
```diff
@@ -29,14 +34,17 @@ public class WorkerTaskRunnerConfig
   private double parallelIndexTaskSlotRatio = 1;
 
+  @JsonProperty
+  private Map<String, @Min(value = 0, message = "Task slot ratio for must be at least 0.") @Max(value = 1, message = "Task slot ratio cannot be greater than 1") Double> taskSlotRatiosPerWorker = new HashMap<>();
```
I'm not sure that adding the constructor would be the best solution here as this is the parent for a couple other classes, and it will require children refactoring as well (including default values specifying), min max annotations seems to be a better approach, if it'll go together with JsonConfigurator update, otherwise I'd better stick with CustomDeserializer validation approach

kfaraz replied:
Hmm, fair point. In that case, let's just skip the validation altogether as there is no validation currently happening for `parallelIndexTaskSlotRatio` either. We can add it when needed later.

nozjkoitop commented on indexing-service/src/test/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfoTest.java:
```diff
@@ -225,11 +239,14 @@ public void test_canRunTask()
         DateTimes.of("2015-01-01T01:01:02Z")
     );
 
+    WorkerTaskRunnerConfig config = mock(WorkerTaskRunnerConfig.class);
```
This will require changing the config class as currently we don't have setters or constructors (except NoArgs), to use actual config object the most painless approach is to add setters and mark them as VisibleForTesting, @kfaraz WDYT?

kfaraz replied:
No, you can also create an inline class like:

```
WorkerTaskRunnerConfig config = new WorkerTaskRunnerConfig() {
    @Override
    public Map<String, Double> getTaskSlotRatiosPerWorker() {
          // return your value here
    }
}
```

Later, we can add a constructor.

nozjkoitop replied:
Thanks for the hint, will do it this way

nozjkoitop replied:
Updated

nozjkoitop commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -160,6 +180,12 @@ public int getCurrParallelIndexCapacityUsed()
   }
 
+  @JsonProperty("currCapacityUsedByTaskType")
+  public Map<String, Integer> getCurrCapacityUsedByTaskType()
```
Done

nozjkoitop commented on processing/src/main/java/org/apache/druid/guice/JsonConfigurator.java:
```diff
@@ -206,6 +206,32 @@ public Message apply(String input)
     return config;
   }
 
+  // To have a detailed error message instead of noSuchFieldException if problematic field is in the superclass
```
Rollbacked

nozjkoitop commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -225,6 +254,43 @@ private int getWorkerParallelIndexCapacity(double parallelIndexTaskSlotRatio)
 
+  private boolean canRunTaskBasedOnLimit(Task task, WorkerTaskRunnerConfig config)
+  {
+    if (task.getType().equals(ParallelIndexSupervisorTask.TYPE) && config.getParallelIndexTaskSlotRatio() != 1) {
```
Nice catch, thanks

nozjkoitop commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -225,6 +254,43 @@ private int getWorkerParallelIndexCapacity(double parallelIndexTaskSlotRatio)
+      return null;
+    }
+
+    return calculateTaskCapacityFromRatio(ratioLimit, worker.getCapacity());
```
Done

nozjkoitop commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -225,6 +254,43 @@ private int getWorkerParallelIndexCapacity(double parallelIndexTaskSlotRatio)
+      Map<String, Double> ratiosMap
+  )
+  {
+    Double ratioLimit = ratiosMap.get(taskType);
```
Renamed

nozjkoitop commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -225,6 +254,43 @@ private int getWorkerParallelIndexCapacity(double parallelIndexTaskSlotRatio)
+                                                                                          .getRequiredCapacity();
+  }
+
+  private Integer getLimitForTask(
```
Renamed

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -225,6 +253,37 @@ private int getWorkerParallelIndexCapacity(double parallelIndexTaskSlotRatio)
+      return true; // No limit specified, so task can run
+    }
+
+    return limit - getCurrCapacityUsedByTaskType().getOrDefault(task.getType(), 0) >= task.getTaskResource()
+                                                                                          .getRequiredCapacity();
```
Formatting:
```suggestion
    return limit - getCurrCapacityUsedByTaskType().getOrDefault(task.getType(), 0)
                       >= task.getTaskResource().getRequiredCapacity();
```

kfaraz reviewed (approved):
LGTM 🚀

kfaraz commented:
@cryptoe , do you have any further comments on this PR?

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -66,6 +71,7 @@ public ImmutableWorkerInfo(
     this.worker = worker;
     this.currCapacityUsed = currCapacityUsed;
     this.currParallelIndexCapacityUsed = currParallelIndexCapacityUsed;
+    this.currCapacityUsedByTaskType = currCapacityUsedByTaskType;
```
```suggestion
    this.currCapacityUsedByTaskType = currCapacityUsedByTaskType == null ? Collections.emptyMap() : currCapacityUsedByTaskType;
```

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -160,6 +180,13 @@ public int getCurrParallelIndexCapacityUsed()
   }
 
+  @JsonProperty("currCapacityUsedByTaskType")
+  @Nullable
```
```suggestion
  @JsonInclude(JsonInclude.Include.NON_EMPTY)
```

kfaraz reviewed (commented):
Making `currCapacityUsedByTaskType` null safe, otherwise several methods would throw NPE.

kfaraz commented:
@nozjkoitop , sorry I had missed these null checks in the earlier review. Could you please include these?
I am unable to commit the suggestions myself.

nozjkoitop commented:
> Could you please include these? I am unable to commit the suggestions myself.

Done, thanks

cryptoe commented on docs/configuration/index.md:
```diff
@@ -1135,6 +1135,7 @@ The following configs only apply if the Overlord is running in remote mode. For
 |`druid.indexer.runner.taskAssignmentTimeout`|How long to wait after a task has been assigned to a Middle Manager before throwing an error.|`PT5M`|
 |`druid.indexer.runner.minWorkerVersion`|The minimum Middle Manager version to send tasks to. The version number is a string. This affects the expected behavior during certain operations like comparison against `druid.worker.version`. Specifically, the version comparison follows dictionary order. Use ISO8601 date format for the version to accommodate date comparisons. |"0"|
 | `druid.indexer.runner.parallelIndexTaskSlotRatio`| The ratio of task slots available for parallel indexing supervisor tasks per worker. The specified value must be in the range `[0, 1]`. |1|
+|`druid.indexer.runner.taskSlotRatiosPerWorker`|Map from task type to the ratio of task slots that the given task type can occupy on a single worker. Each value must be in the range [0, 1]. A value of 0 for a task type means that it can take up no task slot on a worker (i.e. tasks of this type would never run). A value of 1 means that the task type may take up all task slots of a worker if availa ...
```
This might not solve the user problem if its per worker. It has to be global. 
Some implementations do container packing. The approach will fail there.

cryptoe replied:
Please join dev discussion at : https://apachedruidworkspace.slack.com/archives/C030CMF6B70/p1727836829640959

cryptoe commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/autoscaling/PendingTaskBasedWorkerProvisioningStrategy.java:
```diff
@@ -482,10 +482,13 @@ private static ImmutableWorkerInfo workerWithTask(ImmutableWorkerInfo immutableW
         immutableWorker.getWorker(),
         immutableWorker.getCurrCapacityUsed() + 1,
         immutableWorker.getCurrParallelIndexCapacityUsed() + parallelIndexTaskCapacity,
+        immutableWorker.incrementCurrCapacityUsedByTaskType(task.getType(), taskCapacity),
```
Having this per worker is not what end user is expecting. Having a global limit/ratio is what the end user cares about.

kfaraz replied:
In effect, the overall ratio would be more or less the same as every worker would have this same ratio.
`parallelIndexTaskSlotRatio` also does the same thing, so it only makes sense to retain a similar logic.

kfaraz commented:
@nozjkoitop , @cryptoe has expressed some concerns around having a per-worker config

He makes some good points here:
https://apachedruidworkspace.slack.com/archives/C030CMF6B70/p1727836829640959

nozjkoitop commented:
> He makes some good points here: https://apachedruidworkspace.slack.com/archives/C030CMF6B70/p1727836829640959

I'm afraid I don't have access to this Slack workspace

kfaraz commented:
> I'm afraid I don't have access to this Slack workspace

You can try this link [https://druid.apache.org/community/join-slack](https://druid.apache.org/community/join-slack?v=1) from the [Druid docs](https://druid.apache.org/community/).

nozjkoitop commented:
Hey @cryptoe, what do you think about this solution? I've used SelectWorkerStrategies to utilize dynamic configuration of global limits, and it seems like a good option to me.

cryptoe commented:
Apologies, I have been meaning to get to this PR. Will try to finish it by EOW.

nozjkoitop commented:
Hi @cryptoe will you have some time to review this changes?

nozjkoitop commented:
@cryptoe Done

kfaraz commented:
Thanks for your work on this, @nozjkoitop !
I will try to finish the review of this PR later this week.

cryptoe commented on docs/configuration/index.md:
```diff
@@ -1383,6 +1387,16 @@ field. If not provided, the default is to not use it at all.
+|Property|Description|Default|
+|--------|-----------|-------|
+|`taskLimits`|A map where each key is a task type (`String`), and the corresponding value represents the absolute limit on the number of task slots that tasks of this type can occupy. The value is an `Integer` that is greater than or equal to 0. For example, a value of 5 means that tasks of this type can occupy up to 5 task slots in total. If both absolute and ratio limits are specified for the s ...
+|`taskRatios`|A map where each key is a task type (`String`), and the corresponding value is a `Double` which should be in the range [0, 1], representing the ratio of available task slots that tasks of this type can occupy. This ratio defines the proportion of total task slots a task type can use, calculated as `ratio * totalSlots`. If both absolute and ratio limits are specified for the same tas ...
```
```suggestion
|`taskRatios`|A map where each key is a task type (`String`), and the corresponding value is a `Double` which should be in the range [0, 1], representing the ratio of task slots that tasks of this type can occupy. This ratio defines the proportion of total task slots a task type can use, calculated as `ratio * totalSlots`. If both absolute and ratio limits are specified for the same task type, the effective limit will be the smaller of the absolute limit and the limit derived from the corresponding ratio. `taskRatios = {"index_parallel": 0.5, "query_controller": 0.25}`. In this example, parallel indexing tasks can occupy up to 50% of the total task slots, and query controllers can occupy up to 25% of the total task slots.|`{}`|
```

cryptoe commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/setup/LimiterUtils.java:
```diff
@@ -0,0 +1,95 @@
+public class LimiterUtils
+{
+
+  public static boolean canRunTask(Task task, TaskLimits taskLimits, Integer currentSlotsUsed, Integer totalCapacity)
```
This method should have some java docs.

nozjkoitop replied:
Docs added

cryptoe commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/setup/LimiterUtils.java:
```diff
@@ -0,0 +1,95 @@
+
+import java.util.Map;
+
+public class LimiterUtils
```
```suggestion
public class TaskLimiterUtils
```

cryptoe commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -225,6 +244,13 @@ private int getWorkerParallelIndexCapacity(double parallelIndexTaskSlotRatio)
     return workerParallelIndexCapacity;
   }
 
+  public Map<String, Integer> incrementTypeSpecificCapacity(String type, int capacityToAdd)
```
Should there be a corresponding decrement ?

kfaraz replied:
This method is a little misleading given that this is an immutable class.
I would advise simply inlining this change at the caller.

```
updatedMap = new HashMap(immutableWorkerInfo.getCurrCapacityUsedByTaskType()).merge(type, capacityToAdd, Integer::sum);
```

cryptoe commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java:
```diff
@@ -57,6 +61,7 @@ public ImmutableWorkerInfo(
       @JsonProperty("worker") Worker worker,
       @JsonProperty("currCapacityUsed") int currCapacityUsed,
       @JsonProperty("currParallelIndexCapacityUsed") int currParallelIndexCapacityUsed,
+      @JsonProperty("currCapacityUsedByTaskType") Map<String, Integer> currCapacityUsedByTaskType,
```
or this should be immutable no ?

nozjkoitop replied:
It is immutable. This typeSpecificCapacity is managed exactly like parallelIndexCapacityUsed, which is `immutableWorker.getCurrParallelIndexCapacityUsed() + parallelIndexTaskCapacity` in ImmutableWorkerInfo constructor arguments, but here the incremented value is created in ImmutableWorkerInfo itself. Wrt decrement, I dont see any for currParallelIndexCapacityUsed and currCapacityUsed either, if I'm not mistaken it's managed by Provisioning Strategy

cryptoe commented:
Thanks for the changes. 
Lets wait for @kfaraz review and then we can get this merged.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/setup/EqualDistributionWithCategorySpecWorkerSelectStrategy.java:
```diff
@@ -74,7 +85,10 @@ public boolean equals(final Object o)
     }
     final EqualDistributionWithCategorySpecWorkerSelectStrategy that = (EqualDistributionWithCategorySpecWorkerSelectStrategy) o;
-    return Objects.equals(workerCategorySpec, that.workerCategorySpec);
+    if (!Objects.equals(workerCategorySpec, that.workerCategorySpec)) {
+      return false;
+    }
+    return Objects.equals(taskLimits, that.taskLimits);
```
```suggestion
    return Objects.equals(workerCategorySpec, that.workerCategorySpec)
        && Objects.equals(taskLimits, that.taskLimits);
```

kfaraz replied:
Please update the hashCode method as well.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/setup/EqualDistributionWorkerSelectStrategy.java:
```diff
@@ -83,7 +95,10 @@ public boolean equals(final Object o)
     }
     final EqualDistributionWorkerSelectStrategy that = (EqualDistributionWorkerSelectStrategy) o;
-    return Objects.equals(affinityConfig, that.affinityConfig);
+    if (!Objects.equals(affinityConfig, that.affinityConfig)) {
+      return false;
+    }
+    return Objects.equals(taskLimits, that.taskLimits);
```
```suggestion
    return Objects.equals(affinityConfig, that.affinityConfig)
        && Objects.equals(taskLimits, that.taskLimits);
```

kfaraz replied:
Please update the `hashCode` as well.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/setup/FillCapacityWithCategorySpecWorkerSelectStrategy.java:
```diff
@@ -74,7 +86,10 @@ public boolean equals(final Object o)
     }
     final FillCapacityWithCategorySpecWorkerSelectStrategy that = (FillCapacityWithCategorySpecWorkerSelectStrategy) o;
-    return Objects.equals(workerCategorySpec, that.workerCategorySpec);
+    if (!Objects.equals(workerCategorySpec, that.workerCategorySpec)) {
```
fix this and hashCode as suggested for other classes.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/setup/FillCapacityWithCategorySpecWorkerSelectStrategy.java:
```diff
@@ -32,13 +33,17 @@
   )
   {
     this.workerCategorySpec = workerCategorySpec;
+    this.taskLimits = Configs.valueOrDefault(taskLimits, new TaskLimits());
```
Rather than using `new TaskLimits()`, I would advise keeping a `TaskLimits.EMPTY` instance that can be used everywhere.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/setup/FillCapacityWorkerSelectStrategy.java:
```diff
@@ -81,20 +93,26 @@ public boolean equals(final Object o)
   public int hashCode()
   {
-    return Objects.hash(affinityConfig);
+    int result = affinityConfig != null ? affinityConfig.hashCode() : 0;
+    result = 31 * result + (taskLimits != null ? taskLimits.hashCode() : 0);
+    return result;
```
```suggestion
    return Objects.hash(affiinityConfig, taskLimits);
```

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/setup/TaskLimiterUtils.java:
```diff
@@ -0,0 +1,104 @@
+   * @param totalCapacity The total available capacity across all workers.
+   * @return {@code true} if the task meets the defined limits and capacity constraints; {@code false} otherwise.
+   */
+  public static boolean canRunTask(Task task, TaskLimits taskLimits, Integer currentSlotsUsed, Integer totalCapacity)
```
I think this need not be a static utility method.
This could very well live as a non-static method in `TaskLimits` itself.

```
public boolean canRunTask(Task task, Integer currentSlotsUsed, Integer totalCapacity) {
}
```

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/setup/TaskLimiterUtils.java:
```diff
@@ -0,0 +1,104 @@
+
+import java.util.Map;
+
+public class TaskLimiterUtils
```
This class won't be needed if the utility method is moved to `TaskLimits`, where it would be a more natural fit.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/setup/TaskLimits.java:
```diff
@@ -0,0 +1,102 @@
+import java.util.Map;
+import java.util.Objects;
+
+public class TaskLimits
```
Please add a javadoc and keep an `EMPTY` constant instance.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/setup/TaskLimits.java:
```diff
@@ -0,0 +1,102 @@
+
+  public TaskLimits()
+  {
+    this(Collections.emptyMap(), Collections.emptyMap());
```
Since we have now moved to Java 11, you may use this instead for brevity.

```suggestion
    this(Map.of(), Map.of());
```

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/setup/TaskLimits.java:
```diff
@@ -0,0 +1,102 @@
+  private void validateLimits(Map<String, Integer> taskLimits, Map<String, Double> taskRatios)
+  {
+    if (taskLimits != null && taskLimits.values().stream().anyMatch(val -> val < 0)) {
+      throw new IllegalArgumentException("Task limits should be bigger than 0");
```
```suggestion
      throw InvalidInput.exception("Task limit for any type must be greater than zero. Found[%s].", taskLimits);
```

same suggestion for the other exception.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/setup/TaskLimits.java:
```diff
@@ -0,0 +1,102 @@
+  @Override
+  public int hashCode()
+  {
+    int result = taskLimits != null ? taskLimits.hashCode() : 0;
+    result = 31 * result + (taskRatios != null ? taskRatios.hashCode() : 0);
+    return result;
```
Use `Objects.hashCode` for symmetry with `Objects.equals`.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/setup/WorkerSelectUtils.java:
```diff
@@ -145,9 +148,16 @@ public static ImmutableWorkerInfo selectWorker(
+      final TaskLimits taskLimits
   )
   {
+    if (!TaskLimiterUtils.canRunTask(task,
+                                     taskLimits,
+                                     getTotalCapacityUsedByType(allWorkers, task.getType()), getTotalCapacity(allWorkers)
+    )) {
```
style:

```suggestion
    if (
        !TaskLimiterUtils.canRunTask(
            task,
            taskLimits,
            getTotalCapacityUsedByType(allWorkers, task.getType()), getTotalCapacity(allWorkers)
        )
    ) {
```

kfaraz commented on indexing-service/src/test/java/org/apache/druid/indexing/overlord/setup/EqualDistributionWithAffinityWorkerSelectStrategyTest.java:
```diff
@@ -139,12 +144,147 @@ public void testSerde() throws Exception
+                new Worker("http", "localhost0", "localhost0", 2, "v1", WorkerConfig.DEFAULT_CATEGORY), 0,
+                0,
+                capacityUsed,
+                new HashSet<>(),
+                new HashSet<>(),
```
Suggestion:

You may use `Set.of()` instead. Similar, `List.of()` and `Map.of()` are pretty handy too for creating immutable objects.

kfaraz commented on indexing-service/src/test/java/org/apache/druid/indexing/overlord/setup/EqualDistributionWorkerSelectStrategyTest.java:
```diff
@@ -226,7 +226,7 @@ public void testStrongAffinity()
             ),
             true
-        )
+        ), null
```
Please put this in the next line. Same comment above.

kfaraz commented on website/.spelling:
```diff
@@ -2132,6 +2132,7 @@ sqlQuery
 successfulSending
 [S]igar
 taskBlackListCleanupPeriod
+taskLimits
```
You wouldn't need to add it here if you use `taskLimits` in the heading too.

kfaraz replied:
https://github.com/apache/druid/pull/16889#discussion_r1983144494

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/setup/TaskLimits.java:
```diff
@@ -0,0 +1,102 @@
+
+  @JsonCreator
+  public TaskLimits(
+      @JsonProperty("taskLimits") @Nullable Map<String, Integer> taskLimits,
+      @JsonProperty("taskRatios") @Nullable Map<String, Double> taskRatios
```
This is a little confusing, since these two fields are already inside a field named `taskLimits`, right?

I think it would be cleaner if we called these fields something like `maxCountByType`, `maxRatioByType`.

kfaraz reviewed (commented):
Left some suggestions.

kfaraz commented on docs/configuration/index.md:
```diff
@@ -1383,6 +1387,16 @@ field. If not provided, the default is to not use it at all.
 |`categoryMap`|A JSON map object mapping a task type String name to a [CategoryConfig](#categoryconfig) object, by which you can specify category config for different task type.|`{}`|
 |`strong`|With weak workerCategorySpec (the default), tasks for a dataSource may be assigned to other Middle Managers if the Middle Managers specified in `categoryMap` are not able to run all pending tasks in the queue for that dataSource. With strong workerCategorySpec, tasks for a dataSource will only ever be assigned to their specified Middle Managers, and will wait in the pending queue if nec ...
 
+###### taskLimits
```
Use this to avoid spell check failure and adhere with the existing docs style:

```suggestion
###### `taskLimits`
```

nozjkoitop replied:
Thanks, understood

kfaraz reviewed (approved):
Thanks for the changes, @nozjkoitop !
